behaviortree_ros2/include/behaviortree_ros2/bt_topic_sub_node.hpp

```diff
@@ -282,45 +287,93 @@ inline bool RosTopicSubNode<T>::createSubscriber(const std::string& topic_name)
   signal_connection_ = sub_instance_->broadcaster.connect(
       [this](const std::shared_ptr<T> msg) { last_msg_ = msg; });
 
-  topic_name_ = topic_name;
   return true;
 }
+template <class T>
+inline NodeStatus RosTopicSubNode<T>::checkStatus(const NodeStatus& status) const
+{
+  if(!isStatusActive(status))
+  {
+    throw std::logic_error("RosTopicSubNode: the callback must return SUCCESS, FAILURE, "
+                           "or RUNNING");
+  }
+  return status;
+};
 
 template <class T>
-inline NodeStatus RosTopicSubNode<T>::tick()
+inline NodeStatus RosTopicSubNode<T>::onStart()
 {
-  // First, check if the subscriber_ is valid and that the name of the
-  // topic_name in the port didn't change.
-  // otherwise, create a new subscriber
-  std::string topic_name;
-  getInput("topic_name", topic_name);
-
-  if(!topic_name.empty() && topic_name != "__default__placeholder__" &&
-     topic_name != topic_name_)
+  // If the topic name was provided through an input port, is a valid topic name, and is different
+  // from the stored topic name, update the stored topic name to the new string.
+  if(const auto topic_name = getInput<std::string>("topic_name");
+     topic_name.has_value() && !topic_name->empty() &&
+     topic_name.value() != "__default__placeholder__" && topic_name != topic_name_)
   {
-    sub_instance_.reset();
+    topic_name_ = topic_name.value();
   }
 
-  if(!sub_instance_)
+  // If we fail to create a subscriber, exit early and try again on the next tick
+  if(!createSubscriber(topic_name_))
   {
-    createSubscriber(topic_name);
+    return NodeStatus::RUNNING;
   }
 
-  auto CheckStatus = [](NodeStatus status) {
-    if(!isStatusCompleted(status))
-    {
-      throw std::logic_error("RosTopicSubNode: the callback must return"
-                             "either SUCCESS or FAILURE");
-    }
-    return status;
-  };
+  // Discard any previous message unless latching is enabled.
+  // This enforces that the message must have been received after the current call to onStart(), even if this node has been ticked to completion previously.
+  if(!latchLastMessage())
+  {
+    last_msg_.reset();
+  }
+
+  // NOTE(schornakj): Something weird is happening here that prevents the subscriber from receiving a message if the publisher both initializes and publishes a message in between ticks.
+  // I think it has to do with the discovery process between the publisher and subscriber and how that relates to events being handled by the executor.
+  // This might depend on the middleware implemenation too.
+
   sub_instance_->callback_group_executor.spin_some();
-  auto status = CheckStatus(onTick(last_msg_));
+
+  // If no message was received, return RUNNING
+  if(last_msg_ == nullptr)
+  {
+    return NodeStatus::RUNNING;
+  }
+
+  auto status = checkStatus(onTick(last_msg_));
```

Comment:
I think it would be simpler if the user's implementation of `onTick` received the message by const reference instead of as a pointer. It would make it pretty unambiguous that if `onTick` is called then you've received a message, which is a bit friendlier than the current code where you need to check if the pointer actually points to anything before potentially handling the message. I didn't add this here yet because it's a breaking change that would impact all existing code derived from RosTopicSubNode.

Reply:
I think this would prevent the use case where you just want to fail if there is no message, as const ref requires the message to be there.